daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala

```diff
@@ -117,31 +117,33 @@ object Speedy {
       case _ =>
     }
 
-    /** Perform a single step of the machine execution. */
-    def step(): SResult =
-      try {
-        val ctrlToExecute = ctrl
-        // Set control to crash as it must be reset after execution. This guards
-        // against e.g. buggy builtin operations which do not set control and could
-        // then not advance the machine state.
-        ctrl = CtrlCrash(ctrlToExecute)
-        ctrlToExecute.execute(this)
-        SResultContinue
-      } catch {
-        case SpeedyHungry(res: SResult) =>
-          res
-
-        case serr: SError =>
-          serr match {
-            case _: SErrorDamlException if tryHandleException =>
-              SResultContinue
-            case _ => SResultError(serr)
+    /** Run a machine until we get a result: either a final-value of a request for data, with a callback */
+    def run(): SResult = {
+      def again(): SResult = {
+        // note: exception handler is outside while loop
+        try {
+          while (!isFinal) {
+            ctrl.execute(this) // make a single step
           }
-
-        case ex: RuntimeException =>
-          SResultError(SErrorCrash(s"exception: $ex"))
-
+          ctrl match {
+            case CtrlValue(value) => {
+              SResultFinalValue(value) //stop
+            }
+            case _ =>
+              throw SErrorCrash(s"Unexpected ctrl on final machine $ctrl")
+          }
+        } catch {
+          case SpeedyHungry(res: SResult) => res //stop
+          case serr: SError =>
+            serr match {
+              case _: SErrorDamlException if tryHandleException => again() // not tail recursive
```

Comment:
Oh, I missed that call to `again`. We need to avoid recursion here. I don't want to blow through the stack at runtime.

Reply:
fixed